server/src/main/java/org/opensearch/index/IndexSettings.java

```diff
@@ -503,6 +503,20 @@
         Setting.Property.IndexScope
     );
 
+    /**
+     * Expert: sets the amount of time to wait for merges (during {@link org.apache.lucene.index.IndexWriter#commit}
+     * or {@link org.apache.lucene.index.IndexWriter#getReader(boolean, boolean)}) returned by MergePolicy.findFullFlushMerges(...).
+     * If this time is reached, we proceed with the commit based on segments merged up to that point. The merges are not
+     * aborted, and will still run to completion independent of the commit or getReader call, like natural segment merges.
+     */
+    public static final Setting<TimeValue> INDEX_MAX_FULL_FLUSH_MERGE = Setting.timeSetting(
+        "index.max_full_flush_merge",
+        new TimeValue(0, TimeUnit.MILLISECONDS),
```

Comment:
You benchmarked with `10s`; should this be a reasonable default? Or `5s` like the tests? /cc @msokolov

Reply:
@nknize the only heuristic which comes to mind is that it should not be larger than `index.refresh_interval`: may be we could cap it to be max(half of the refresh interval, 10s) by default (if not specified), wdyt?

Reply:
I think that's a reasonable default. Let's add that to this PR.

Reply:
👍 will do that shortly

Reply:
The original idea was to avoid creating lots of tiny segments while not interfering with the normal merge policy merges or other operations. The idea of the timeout was that we didn't want to block other operations for too long, since this merge-on-commit or merge-on-refresh is *blocking* the corresponding operation from completing. I don't know what refresh interval typically is set to - 1 minute maybe? That would lead to default of 10s. We've tested this with 30, and it works ok for our indexing workload, but YMMV. 10s seems safe to me

Reply:
:+1: thanks a lot @msokolov , 10s it be @nknize ?